### PR TITLE
fixes for 2 possible sources of crashes

### PR DESCRIPTION
**Pull request by msspad**

### CODE REVIEW

File: NanoVNASaver/Windows/TDR.py [hunk 1]
```diff
@@ -34,7 +34,7 @@ def __init__(self, app: QtWidgets.QWidget):
         super().__init__()
         self.app = app
 
-        self.td = []
+        self.td = None
```
Comment (zarath) [inline NanoVNASaver/Windows/TDR.py:37]:
```suggestion
        self.td = []
```
I prefer variable initialization with target type over "None".
Both will evaluate to false in if statement. But if you do something like
for elem in self.td:
   ....
None is a source of error.

File: NanoVNASaver/Hardware/NanoVNA_V2.py [hunk 2]
```diff
@@ -223,21 +223,6 @@ def readVersion(self) -> 'Version':
         logger.debug("readVersion: %s", result)
         return result
 
-    def read_board_revision(self) -> 'Version':
```
Comment (zarath) [inline NanoVNASaver/Hardware/NanoVNA_V2.py:226]: This isn't a duplicate of readVersion.

readVerstion calls CMD FIRMWARE_MAJOR and CMD FIRMWARE_MINOR (0xf3, 0xf4)
read_board_revicsion calls ADDR_DEVICE_VARIANT and ADDR_HARDWARE_REVISION (0xf1, 0xf2)

File: NanoVNASaver/Hardware/NanoVNA_V2.py [hunk 3]
```diff
@@ -223,21 +223,6 @@ def readVersion(self) -> 'Version':
...
-        cmd = pack("<BBBB",
-                   _CMD_READ, _ADDR_DEVICE_VARIANT,
-                   _CMD_READ, _ADDR_HARDWARE_REVISION)
-        with self.serial.lock:
-            self.serial.write(cmd)
-            sleep(WRITE_SLEEP)
-            resp = self.serial.read(2)
-        if len(resp) != 2:
-            logger.error("Timeout reading version registers")
-            return None
```
Comment (zarath) [inline NanoVNASaver/Hardware/NanoVNA_V2.py:236]: raise IOError here to be consistent with readVersion's behavior!

**Comment by zarath** [comment]
Thank you for your request.
I've commented some parts.
Can you check this?

Regards
   Holger

File: NanoVNASaver/Windows/TDR.py (hunk 1 above)
Comment (msspad) [reply NanoVNASaver/Windows/TDR.py:37]: it's fine to initialize empty objects. However, td is at least sometimes a numpy array and not a list. That means e.g. `if self.tdrWindow.td` raises a ValueError when td has 2 elements and more:
```
if np.empty(0):
	print("empty arrays evaluate to true but raise a DeprecationWarning and will raise an error in the future")
if np.zeros((1,)):
	print("unfortunately, this just passes?")
if np.zeros((2,)):
	print("arrays with more than 1 element raise a ValueError, so I get never printed")
```

You can `self.td = np.empty(0)` and then check if td is non-empty:
```
td = np.zeros((2,))
if td.size != 0:
	print("success!")
```

File: NanoVNASaver/Hardware/NanoVNA_V2.py (hunk 2 above)
Comment (msspad) [reply NanoVNASaver/Hardware/NanoVNA_V2.py:226]: I missed that, thanks. Then you can just add the IOError to read_board_revision() to fix the error, like you suggested below.

**Comment by zarath** [comment]
@msspad I have changed the TDRWindow to initialize td to an empty numpy array and Chart to use td.size in all if statements,
but can only test it tomorrow as I haven't the device at hand.
Can you supply that type annotation patch as PR to get mentioned in git authorship?

**Comment by msspad** [comment]
Happy to provide the PR, probably no sooner than Sunday. I don't care about the git authorship, so if you want to go ahead without a PR, I'm totally fine. I only care about great tools, and nanovna-saver is definitely a great one! Thanks for your fantastic work, msspad

**Pull request closed without merging**